File: solutions/python/satellite/1/satellite.py

def tree_from_trainsversals_recursive(preorder, inorder):
    if len(preorder) == 0:
        return {}
    
    v = preorder[0]
    index_v = inorder.index(v)

    new_left_inorder = inorder[:index_v]
    new_left_preorder = [ item for item in preorder if item in new_left_inorder ]
    l = tree_from_trainsversals_recursive(new_left_preorder, new_left_inorder)

    new_right_inorder = inorder[(index_v + 1):]
    new_right_preorder = [ item for item in preorder if item in new_right_inorder ]
    r = tree_from_trainsversals_recursive(new_right_preorder, new_right_inorder)

    return { "v": v, "l": l, "r": r}


def tree_from_traversals(preorder, inorder):
    len_preorder = len(preorder)
    len_inorder = len(inorder)
    if len_preorder != len_inorder:
        raise ValueError("traversals must have the same length")
    if set(preorder) != set(inorder):
        raise ValueError("traversals must have the same elements")
    if len(set(preorder)) < len(preorder) or len(set(inorder)) < len(inorder):
        raise ValueError("traversals must contain unique items")
    if len_inorder == 0:
        return {}
    
    return tree_from_trainsversals_recursive(preorder, inorder)

File: solutions/java/satellite/1/src/main/java/Satellite.java
import java.util.List;
import java.util.Set;

public class Satellite {


    private Node nodeFromTransversals(List<Character> preorderInput, List<Character> inorderInput) {

        int sizeInorder = inorderInput.size();
        if (sizeInorder == 0) {
            return null;
        }

        Character value = preorderInput.get(0);
        int indexValue = inorderInput.indexOf(value);

        List<Character> leftInorder = inorderInput.subList(0, indexValue);
        List<Character> leftPreorder = preorderInput.stream().filter(leftInorder::contains).toList();
        Node leftSubtree = nodeFromTransversals(leftPreorder, leftInorder);

        List<Character> rightInorder = inorderInput.subList(indexValue + 1, sizeInorder);
        List<Character> rightPreorder = preorderInput.stream().filter(rightInorder::contains).toList();
        Node rightSubtree = nodeFromTransversals(rightPreorder, rightInorder);

        Node root = new Node(value);
        root.left = leftSubtree;
        root.right = rightSubtree; 

        return root;
    }

    public Tree treeFromTraversals(List<Character> preorderInput, List<Character> inorderInput) {
        
        int sizePreorder = preorderInput.size();
        int sizeInorder = inorderInput.size();

        if (sizePreorder != sizeInorder) {
            throw new IllegalArgumentException("traversals must have the same length");
        }

        Set<Character> preorderSet = Set.copyOf(preorderInput);
        Set<Character> inorderSet = Set.copyOf(inorderInput);

        if (!preorderSet.equals(inorderSet)) {
            throw new IllegalArgumentException("traversals must have the same elements");
        }

        if (preorderSet.size() != sizePreorder || inorderSet.size() != sizeInorder) {
            throw new IllegalArgumentException("traversals must contain unique items");
        }

        return new Tree(nodeFromTransversals(preorderInput, inorderInput));
    }
}
